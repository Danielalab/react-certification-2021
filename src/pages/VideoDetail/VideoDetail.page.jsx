import React from 'react';
import styled from 'styled-components';
import { useParams } from 'react-router';
import VideosList from '../../components/VideosList';
import VideoPlayer from '../../components/VideoPlayer';
import useSearchQuery from '../../hooks/useSearchQuery';
import useVideosQuery from '../../hooks/useVideosQuery';

const Wrapper = styled.section`
  display: flex;
  flex-wrap: wrap;
  justify-content: space-between;
  padding: 2rem;
  margin-top: 70px;
  @media (max-width: 576px) {
    justify-content: center;
    padding: 1rem;
  }
`;

const VideoPlayerWrapper = styled.div`
  width: 70%;
  @media (max-width: 576px) {
    width: 100%;
  }
`;

function VideoDetailPage() {
  const { id } = useParams();
  const [data, loading, error] = useVideosQuery({ videoId: id });
  const [relatedVideos, loadingRelatedVideos] = useSearchQuery({
    relatedToVideoId: id,
  });
  const video = ((data || {}).items || [])[0];
  return (
    <Wrapper>
      {loading && '...loading'}
      {error && 'Something was wrong'}
      {data && !loading && (
        <VideoPlayerWrapper>
          <VideoPlayer
            videoId={video.id}
            title={video.snippet.title}
            description={video.snippet.description}
            publishedAt={video.snippet.publishedAt}
          />
        </VideoPlayerWrapper>
      )}
      {relatedVideos && !loadingRelatedVideos && (
        <div>
          <VideosList videos={relatedVideos.items} />
        </div>
      )}
    </Wrapper>
  );
}

export default VideoDetailPage;
