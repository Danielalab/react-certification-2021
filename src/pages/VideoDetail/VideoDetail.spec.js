import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import VideoDetail from './VideoDetail.page';
import dataDummy from '../../hooks/tests/responses/searchQuerySuccess.json';

const { render, screen, waitFor } = require('@testing-library/react');

/* mocking children componnets */
jest.mock('../../components/VideoThumbnail/VideoThumbnail.component');
jest.mock('../../components/VideoPlayer/VideoPlayer.component');

describe('VideoDetail Page', () => {
  test('Should renders VideoDetail page component with VideoPlayer & VideosList', async () => {
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.resolve(dataDummy),
      })
    );
    /* render component with router */
    const history = createMemoryHistory();
    const { container } = render(
      <Router history={history}>
        <VideoDetail />
      </Router>
    );
    history.push('/fakeVideoId1');

    let loading;
    let videosListsComponent;
    let videoPlayerComponent;

    loading = screen.getByText(/...loading/i);
    videosListsComponent = screen.queryByText('videos-list');
    videoPlayerComponent = screen.queryByText('video-player');
    expect(container).toBeInTheDocument();
    expect(loading).toBeInTheDocument();
    expect(videosListsComponent).not.toBeInTheDocument();
    expect(videoPlayerComponent).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();

    await waitFor(() => {
      screen.getAllByTestId('video-thumbnail');
      expect(screen.getByTestId('video-player')).toBeInTheDocument();
    });
    videosListsComponent = screen.getAllByTestId('video-thumbnail');
    videoPlayerComponent = screen.getByTestId('video-player');
    loading = screen.queryByText(/...loading/i);
    expect(videosListsComponent[0]).toBeInTheDocument();
    expect(videosListsComponent[0].textContent).toBe(
      'Dummy Video Thumbnail component fakeVideoId1'
    );
    expect(videoPlayerComponent).toBeInTheDocument();
    expect(videoPlayerComponent.textContent).toBe('Dummy Video Player component');
    expect(loading).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  test('Should renders an Error message', async () => {
    let loading;
    const history = createMemoryHistory();
    global.fetch = jest.fn(() =>
      Promise.resolve({
        json: () => Promise.reject(new Error('API error')),
      })
    );

    /* render component with router */
    const { container } = render(
      <Router history={history}>
        <VideoDetail />
      </Router>
    );
    history.push('/fakeVideoId1');

    loading = screen.getByText(/...loading/i);
    expect(container).toBeInTheDocument();
    expect(loading).toBeInTheDocument();
    expect(container).toMatchSnapshot();

    await waitFor(() => screen.getByText(/Something was wrong/i));
    const errorMessage = screen.getByText(/Something was wrong/i);
    expect(errorMessage).toBeInTheDocument();
    loading = screen.queryByText(/...loading/i);
    expect(loading).not.toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });
});
