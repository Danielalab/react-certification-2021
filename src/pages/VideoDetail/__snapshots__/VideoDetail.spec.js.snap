// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideoDetail Page Should renders VideoDetail page component with VideoPlayer & VideosList 1`] = `
<div>
  <section
    class="sc-bdnxRM hyfyNv"
  >
    ...loading
  </section>
</div>
`;

exports[`VideoDetail Page Should renders VideoDetail page component with VideoPlayer & VideosList 2`] = `
<div>
  <section
    class="sc-bdnxRM hyfyNv"
  >
    <div
      class="sc-gtsrHT csNJBk"
    >
      <div
        data-testid="video-player"
      >
        <p>
          Dummy Video Player component
        </p>
      </div>
    </div>
    <div>
      <div
        data-testid="videos-list"
      >
        <p>
          Dummy Videos List component
        </p>
      </div>
    </div>
  </section>
</div>
`;

exports[`VideoDetail Page Should renders an Error message 1`] = `
<div>
  <section
    class="sc-bdnxRM hyfyNv"
  >
    ...loading
  </section>
</div>
`;

exports[`VideoDetail Page Should renders an Error message 2`] = `
<div>
  <section
    class="sc-bdnxRM hyfyNv"
  >
    Something was wrong
  </section>
</div>
`;
